File: java/src/test/java/org/sammancoaching/testing/HttpServerLifecycle.java
package org.sammancoaching.testing;

import org.junit.jupiter.api.extension.AfterEachCallback;
import org.junit.jupiter.api.extension.BeforeEachCallback;
import org.junit.jupiter.api.extension.ExtensionContext;

/**
 * JUnit 5 extension for managing an HttpServer lifecycle in tests.
 * Starts a lightweight HTTP server before the test
 * and stores a reference to it in a thread local variable where the test can retrieve it.
 * Stops the server after the test.
 */
public class HttpServerLifecycle implements BeforeEachCallback, AfterEachCallback {

    private static final String DEFAULT_HOST = "127.0.0.1";
    private static final int DEFAULT_PORT = 9998;
    
    private static LightweightHttpServer currentHttpServer;
    
    private final String host;
    private final int port;
    
    public HttpServerLifecycle() {
        this(DEFAULT_HOST, DEFAULT_PORT);
    }
    
    public HttpServerLifecycle(String host, int port) {
        this.host = host;
        this.port = port;
    }
    
    @Override
    public void beforeEach(ExtensionContext context) throws Exception {
        LightweightHttpServer server = new LightweightHttpServer(host, port);
        server.start();

        currentHttpServer = server;
    }
    
    @Override
    public void afterEach(ExtensionContext context) throws Exception {
        if (currentHttpServer != null) {
            currentHttpServer.stop();
            currentHttpServer.getHandler().reset();
        }
        currentHttpServer = null;
    }
    
    public static LightweightHttpServer getHttpServer() {
        return currentHttpServer;
    }

}

File: kotlin/src/test/kotlin/org/sammancoaching/testing/HttpServerLifecycle.kt
package org.sammancoaching.testing

import org.junit.jupiter.api.extension.AfterEachCallback
import org.junit.jupiter.api.extension.BeforeEachCallback
import org.junit.jupiter.api.extension.ExtensionContext


/**
 * JUnit 5 extension for managing an HttpServer lifecycle in tests.
 * Starts a lightweight HTTP server before the test
 * and stores a reference to it in a field where the test can retrieve it.
 * Stops the server after the test.
 */
class HttpServerLifecycle @JvmOverloads constructor(
    private val host: String = DEFAULT_HOST,
    private val port: Int = DEFAULT_PORT
) : BeforeEachCallback, AfterEachCallback {
    @Throws(Exception::class)
    override fun beforeEach(context: ExtensionContext?) {
        val server = LightweightHttpServer(host, port)
        server.start()

        httpServer = server
    }

    @Throws(Exception::class)
    override fun afterEach(context: ExtensionContext?) {
        if (httpServer != null) {
            httpServer!!.stop()
            httpServer!!.getHandler().reset()
        }
        httpServer = null
    }

    companion object {
        private const val DEFAULT_HOST = "127.0.0.1"
        private const val DEFAULT_PORT = 9998

        var httpServer: LightweightHttpServer? = null
            private set
    }
}